### PR TITLE
[BugFix] Fix pk concurrent apply issue

### PR DESCRIPTION
**Pull request by sevev**
## Why I'm doing:
There's an issue with pk tables applying at the same time:
1. BE receive a increment clone request, and tablet1(tablet id is x) execute apply in thread1
2. BE receive drop tablet1 request and erase it from `tablet_manager`,  but drop is not finished because need to wait for apply finished.
3. BE receive a new clone request and create a new tablet2(tablet id is also x).
4. BE receive a new incremental clone request and tablet2 execute apply in thread2
5. tablet1 and tablet2 are doing apply at the same time. These two tablets are different tablet but has the same tablet id, so they will get the same `primary index` in memory and update at the same time. It will cause unexpected issue.

## What I'm doing:
1. remove tablet from `tablet_manager` at the last step of dropping tablet.
2. reject clone request when we find tablet is under dropping but not finished.

Fixes #issue

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/storage/tablet.h [hunk 1]
```diff
@@ -450,6 +454,8 @@ class Tablet : public BaseTablet {
     // another tablet with the same tablet id
     // currently, it will be used in Restore process
     bool _will_be_force_replaced = false;
+
+    std::atomic<bool> _is_dropping{false};
```
Comment (luohaha) [inline be/src/storage/tablet.h:458]: Is it necessary to add a new state?

Comment (sevev) [reply be/src/storage/tablet.h:458]: > Is it necessary to add a new state?

When drop and clone operations are concurrent, different execution sequences can lead to various outcomes, such as the clone succeed while the apply stop. In fact, during the drop tablet process, once the drop operation begins, any subsequent clone requests on this tablet are meaningless. Therefore, directly rejecting these requests would be better and can also avoid complex concurrency issues.

File: be/src/storage/tablet_manager.cpp [hunk 2]
```diff
@@ -436,6 +439,16 @@ Status TabletManager::drop_tablet(TTabletId tablet_id, TabletDropFlag flag) {
...
     }
+    // erase tablet from tablet map
+    {
+        std::unique_lock wlock(_get_tablets_shard_lock(tablet_id));
+        TabletMap& tablet_map = _get_tablet_map(tablet_id);
+        auto it = tablet_map.find(tablet_id);
+        if (it != tablet_map.end()) {
+            tablet_map.erase(it);
+        }
+        _remove_tablet_from_partition(*dropped_tablet);
```
Comment (trueeyu) [inline be/src/storage/tablet_manager.cpp:450]: If not found in map, alse `_remove_tablet_from_partition`?

**Comment by trueeyu** [comment]
` These two tablets are different tablet but has the same tablet id`

->

`. These two tablet instances of one tablet  has the same tablet id`

File: be/src/storage/tablet_manager.cpp [hunk 3]
```diff
@@ -394,8 +394,11 @@ Status TabletManager::drop_tablet(TTabletId tablet_id, TabletDropFlag flag) {
 
         LOG(INFO) << "Start to drop tablet " << tablet_id;
         dropped_tablet = it->second;
-        tablet_map.erase(it);
-        _remove_tablet_from_partition(*dropped_tablet);
+        dropped_tablet->set_is_dropping(true);
```
Comment (trueeyu) [inline be/src/storage/tablet_manager.cpp:397]: Without acquiring the Tablet lock, can the scope of this lock completely solve the concurrency problem?

**Review by trueeyu — APPROVED** [review]

**Review by luohaha — APPROVED** [review]

**Review by wyb — APPROVED** [review]

**Pull request merged**